### PR TITLE
rustc_codegen_ssa: workaround broken verbatim behavior

### PR DESCRIPTION
**Pull request by liushuyu**
This pull request workaround broken verbatim behaviours for several GNU-flavored linkers.

You'd be surprised how this simple mechanism can be implemented so ~~botched~~ differently across many GNU-compatible linkers.

According to my testing:

| Linker | `-l:/path/to/library` behavior |
|-----------|------------------------------------------|
| `ld.bfd` | :x:  Appends `/path/to/library` to the end of each search path, e.g. `/usr/lib//path/to/library`. |
| `ld.gold` | :heavy_check_mark:  Working as expected. |
| `ld.lld` | :heavy_check_mark:  Working as expected. |
| `ld.mold` | :x: Can't find any library called `:/path/to/library`. |

This pull request uses a similar strategy to CMake, where we just pass the library name directly to the linker if we detect that `ld.lld` is not used.

This may have the side effect that the user can pass arbitrary linker flags (including linker scripts) to the linker through the `+verbatim` modifier (you could argue this is what "verbatim" means).

### CODE REVIEW

**Comment by workingjubilee** [comment]
...isn't working around their handling the precise opposite of passing it verbatim?

**Comment by workingjubilee** [comment]
I suppose one could level that accusation at our current behavior, in a way, since we still have to conditionalize it on the linker's supposed API surface, so I'm not saying we shouldn't do this, I'm just curious.

**Comment by liushuyu** [comment]
> ...isn't working around their handling the precise opposite of passing it verbatim?

Well... in a way, yes. But I suppose passing the library as `-l:<...>`  to a linker that does not support this syntax is worse?

**Comment by liushuyu** [comment]
> I suppose one could level that accusation at our current behavior, in a way, since we still have to conditionalize it on the linker's supposed API surface, so I'm not saying we shouldn't do this, I'm just curious.

Unfortunately, this is another GNU-extended behaviour that everyone (including GNU binutils) implemented badly. So, "passing library path verbatim" for different linkers needs to be handled... differently, like what we already have done for other linker-related shenanigans.

**Comment by workingjubilee** [comment]
perhaps we need to coin a new word to describe this curious behavior? `+gnubatim`

**Comment by liushuyu** [comment]
> perhaps we need to coin a new word to describe this curious behavior? `+gnubatim`

Haha, maybe. But the real trouble here is that we don't really identify the linker as CMake does. Otherwise, we can just exclude the problematic ones and call it a day.

Another issue I can see here is how "verbatim" we should allow this option to be. For example, should we escape the library names so that a file called `foobar@version` may not have some other meaning to the linker? Or if a library's path name begins with `-` for some unknown reason?

**Comment by jieyouxu** [comment]
I'm rerolling this PR preemptively to another compiler reviewer because I'm completely :ferrisClueless: about weird linker shenanigans.

r? @cjgillot (random guess that you might feel more comfortable reviewing this, please feel free to reroll)

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -1497,12 +1497,18 @@ fn print_native_static_libs(
...
+                            Some(format!("-Wl,{}", name))
+                        }
+                        LinkerFlavor::Gnu(Cc::No, Lld::No) if verbatim => Some(name.to_string()),
+                        LinkerFlavor::Gnu(..) => {
+                            Some(format!("-l{}{}", if verbatim { ":" } else { "" }, name))
+                        }
+                        _ if sess.target.is_like_msvc => {
+                            Some(format!("{}{}", name, if verbatim { "" } else { ".lib" }))
+                        }
+                        _ => Some(format!("-l{name}")),
```
Comment (jieyouxu) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1511]: Discussion: this seems... very tricky.

(This question is not necessarily directed at you) Are we even able to test for this in CI, as in picking *different* linkers?

Comment (jieyouxu) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1511]: Yeah, that's what I was thinking, it's a matter of if we even have these linkers in CI installed.

**Comment by saethlin** [comment]
> You'd be surprised how this simple mechanism can be implemented so ~botched~ differently across many GNU-compatible linkers.

I'm not surprised but I'm still disappointed GNU being incompatible with itself.

> `ld.mold`	❌ Can't find any library called `:/path/to/library`.

Isn't this just a bug in mold? Is there an issue filed for it?

**Comment by liushuyu** [comment]
> > You'd be surprised how this simple mechanism can be implemented so ~botched~ differently across many GNU-compatible linkers.
> 
> I'm not surprised but I'm still disappointed GNU being incompatible with itself.
> 
> > `ld.mold`	❌ Can't find any library called `:/path/to/library`.
> 
> Isn't this just a bug in mold? Is there an issue filed for it?

I am unsure if this is a bug because, upon some more digging, it seems like `mold` just prints out what you gave it on the command line. It does correctly strip the leading `:`, though.

Anyways, the problem I encountered was that if you supply the linker with an absolute path to the static library via `-l:/path/to/library`, different linkers will react differently.

I am also unsure if the `+verbatim` modifier in `rustc` should support this use case. Some issues could be stemming from https://github.com/rust-lang/rust/blob/e35364a521372ce682e4bd4a5850d97ea33b0eab/library/profiler_builtins/build.rs#L11 (where the downstream distributor could supply an absolute path to a prebuilt LLVM artifact).

**Comment by petrochenkov** [comment]
The `ld.bfd` behavior is the canonical expected behavior here, see `namespec` in https://sourceware.org/binutils/docs/ld/Options.html

`-l:` is just not supposed to be used with full paths.
It is only supposed to be used to write
- `-l:libfoo.a` for disambiguation if favor of a static lib, in Rust it's preferable to use library kinds for this
- `-l:non_libfoo.non_a` for custom prefixes and suffixes

If the standard `ld.bfd` `-l:`semantics don't work for you, then `+verbatim` shouldn't be used as well (for gnu-style linkers at least).
Something else should be used - library kinds, full paths, etc.

**Comment by liushuyu** [comment]
> The `ld.bfd` behavior is the canonical expected behavior here, see `namespec` in https://sourceware.org/binutils/docs/ld/Options.html

Okay, so the documentation is a bit confusing but `ld.bfd` sticks to the correct behavior.



> `-l:` is just not supposed to be used with full paths. It is only supposed to be used to write
> 
>     * `-l:libfoo.a` for disambiguation if favor of a static lib, in Rust it's preferable to use library kinds for this
> 
>     * `-l:non_libfoo.non_a` for custom prefixes and suffixes
> 
> 
> If the standard `ld.bfd` `-l:`semantics don't work for you, then `+verbatim` shouldn't be used as well (for gnu-style linkers at least). Something else should be used - library kinds, full paths, etc.
> 
> Upd: `-l` without the colon is not supposed to be used with full paths too.

Understood. I could whip up a workaround in this case, however, I do wonder if this kind of behavior is expected in `rustc` (not whatever linker should be doing)?

**Comment by JohnCSimon** [comment]
Ping from triage: 
@liushuyu can you post the status of this PR?

**Comment by liushuyu** [comment]
> Ping from triage: @liushuyu can you post the status of this PR?

I think we can close this pull request, due to this is an expected behavior.

**Pull request closed without merging**